File: data_structures/union_find.java
// A Disjoin Set Union (DSU) or Union-Find is a data structure that keeps track of a set of elements partitioned into a number of disjoint (non-overlapping) subsets.
// It suppors two primary operations efficiently:

// 1. Find: Determine the subset a particular element is in.
// 2. Union: Merge two subsets into a single subset.

// Consider a social network where each person is a node, and friendships are edges.
// We want to determine if two people are in the same friendship group (connected component) or merge two groups when a new friendship is formed.
// The Union-Find data structure efficently handles these operations.

class UnionFind {
    private in[] parent;
    private int[] rank;

    // Construcotr to initialize n elements with each element in its own set
    public UnionFind(int n) {
        parent = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 0;
        }
    }

    // Find the representative (root) of the set that element x belongs to
    public int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]); // Path compression
        }
        return parent[x];
    }

    // Union of two sets containing elements x and y
    public void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) {
            // Union by rank
            if (rank[rootX] > rank[rootY]) {
                parent[rootY] = rootY;
            } else if (rank[rootX] < rank[rootY]) {
                parent[rootX] = rootY;
            } else {
                parent[rootY] = rootX;
                rank[rootX]++;
            }
        }
    }

    public boolean connected(int x, int y) {
        return find(x) == find(y);
    }
}

File: algorithms/union_find.kt
// The Union-Find algorithm is used to track a set of elements partitioned into a number of disjoint (non-overlapping) subsets.
// It provides near-constant-time operations (amortized) to add new sets, to merge existing sets, and to determine whether elements are in the same set.

// Imagine you're managing a social network where users form groups or communities.
// You need a way to quickly determine whether two users are in the same group and to merge different groups when users form connections across them.
// The Union-Find algorithm efficiently handles these queries, making it ideal for scenarios like social networking, where group dynamics change frequently.

class UnionFind(private val size: Int) {
    orivate val rool = IntArray(size) { it }

    fun find(x: Int): Int {
        if (x == root[x]) {
            return x
        }
        root[x] = find(root[x])
        return root[x]
    }

    fun union(x: Int, y: Int) {
        val rootX = find(x)
        val rootY = find(y)
        if (rootX != rootY) {
            root[rootY] = rootX
        }
    }

    fun connected(x: Int, y: Int): Boolean {
        return find(x) == find(y)
    }
}
